### PR TITLE
HYDRATOR-1426 support setting pipeline properties

### PR DESCRIPTION
**Pull request by albertshau**
Pipeline properties can now be set in the application config.
If prefixed by 'system.[engine].', the property will be set
in the corresponding Configuration for mapreduce or SparkConf
for spark. Plugins can also set properties at configure time,
with some checks to make sure multiple plugins don't set different
values for the same property key.

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT5652

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Engine.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+package co.cask.cdap.etl.api;
+
+/**
+ * Execution engines.
+ */
+public enum Engine {
+  MAPREDUCE,
+  SPARK,
+  // deprecated engines
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Engine.java]: We can annotate this with @Deprecated

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/PipelineConfigurer.java [hunk 2]
```diff
@@ -31,4 +33,18 @@
...
   StageConfigurer getStageConfigurer();
+
+  /**
+   * @return the engine for this pipeline
+   */
+  Engine getEngine();
+
+  /**
+   * Set pipeline properties that will be applied to each run of the pipeline.
+   * Depending on the engine used, the properties will be added to the SparkConf for each run or the mapreduce
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/PipelineConfigurer.java:44]: are all properties added to the config or only those which are prefixed `system.spark`/`system.mapreduce`?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/PipelineSpec.java [hunk 3]
```diff
@@ -197,9 +212,15 @@ public T setNumOfRecordsPreview(int numOfRecordsPreview) {
       return (T) this;
     }
 
+    public T setProperties(Map<String, String> properties) {
+      this.properties.clear();;
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/PipelineSpec.java]: nit: remove extra `;`

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/PipelineSpec.java [hunk 4]
```diff
@@ -120,6 +132,7 @@ public String toString() {
       ", clientResources=" + clientResources +
       ", stageLoggingEnabled=" + stageLoggingEnabled +
       ", numOfRecordsPreview=" + numOfRecordsPreview +
+      ", properties=" + properties +
       '}';
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/spec/PipelineSpec.java]: nit: `"}";`

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Engine.java (hunk 1 above)
Comment (sagarkapare) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Engine.java]: Also its bit weird to add new deprecated enum constant.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/PipelineConfigurer.java (hunk 2 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/PipelineConfigurer.java:44]: only those that are prefixed. This is meant to mirror whatever we would eventually do with runtime args in cdap, to set things in spark conf for user programs.

Comment (sagarkapare) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/PipelineConfigurer.java:44]: should we update the javadoc to reflect that?

**Comment by sagarkapare** [comment]
one comment about updating the javadoc. rest LGTM 👍

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/PipelineConfigurer.java (hunk 2 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/PipelineConfigurer.java:44]: oh sorry I thought this was somewhere else. When set programmatically, no prefixing is needed. The prefixing is only used when somebody sets the properties through the application config.

**Pull request merged**